File: src/main/java/com/nanaios/polygonal_tech/capability/impl/LongEnergyStorage.java
package com.nanaios.polygonal_tech.capability.impl;

import com.nanaios.polygonal_tech.capability.interfaces.ILongEnergyStorage;

import java.util.function.BooleanSupplier;
import java.util.function.LongSupplier;

public class LongEnergyStorage implements ILongEnergyStorage {
    private long energy;
    private final LongSupplier capacity;
    private final BooleanSupplier canReceive;
    private final BooleanSupplier canExtract;

    public LongEnergyStorage(
            LongSupplier capacity,
            BooleanSupplier canReceive,
            BooleanSupplier canExtract
    ) {
        this.capacity = capacity;
        this.canReceive = canReceive;
        this.canExtract = canExtract;
    }

    @Override
    public long getLongEnergyStored() {
        return energy;
    }

    @Override
    public long getLongMaxEnergyStored() {
        return capacity.getAsLong();
    }

    @Override
    public long receiveLongEnergy(long maxReceive, boolean simulate) {
        if (!canReceive())
            return 0;

        long energyReceived = Math.min(capacity.getAsLong() - energy, maxReceive);
        if (!simulate)
            energy += energyReceived;
        return energyReceived;
    }

    @Override
    public long extractLongEnergy(long maxExtract, boolean simulate) {
        if (!canExtract())
            return 0;

        long energyExtracted = Math.min(energy, maxExtract);
        if (!simulate)
            energy -= energyExtracted;
        return energyExtracted;
    }

    @Override
    public boolean canExtract() {
        return canExtract.getAsBoolean();
    }

    @Override
    public boolean canReceive() {
        return canReceive.getAsBoolean();
    }

    @Override
    public boolean canOutput() {
        return canExtract.getAsBoolean();
    }

    @Override
    public boolean canInput() {
        return canReceive.getAsBoolean();
    }
}

File: src/main/kotlin/com/nanaios/polygonal_tech/core/capability/energy/LongEnergyStorage.kt
package com.nanaios.polygonal_tech.core.capability.energy

import com.nanaios.polygonal_tech.core.capability.energy.ILongEnergyStorage.Companion.NBT_CAPACITY_KEY
import com.nanaios.polygonal_tech.core.capability.energy.ILongEnergyStorage.Companion.NBT_STORED_KEY
import com.nanaios.polygonal_tech.core.network.sync.storage.EmptySyncValueStorage
import com.nanaios.polygonal_tech.core.network.sync.storage.ISyncValueStorage
import com.nanaios.polygonal_tech.core.network.sync.type.ISyncType
import com.nanaios.polygonal_tech.core.network.sync.type.SyncType
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.FriendlyByteBuf
import kotlin.math.min

/**
 * [ILongEnergyStorage]の標準的な実装クラスであり、
 * NBTへの永続化、ネットワークへのバイトバッファ送受信機能（[ISyncValue]）を内包することを目的とする。
 *
 * @param stored 初期状態で保持するエネルギー量
 * @param capacity 蓄電可能な最大エネルギー容量
 */
open class LongEnergyStorage(
    stored: Long,
    capacity: Long,
) : ILongEnergyStorage {
    override var longEnergyStored = stored
        protected set
    override var maxLongEnergyStored = capacity
        protected set
    override var type: ISyncType = SyncType.NONE
    override var storage: ISyncValueStorage = EmptySyncValueStorage
    override var isDirty = false
        protected set

    protected fun markDirty() {
        isDirty = true
        storage.onSyncValueChanged(this)
    }

    override fun serializeNBT(): CompoundTag {
        val tag = CompoundTag()
        tag.putLong(NBT_STORED_KEY, this.longEnergyStored)
        tag.putLong(NBT_CAPACITY_KEY, this.maxLongEnergyStored)
        return tag
    }

    override fun deserializeNBT(nbt: CompoundTag) {
        this.longEnergyStored = nbt.getLong(NBT_STORED_KEY)
        this.maxLongEnergyStored = nbt.getLong(NBT_CAPACITY_KEY)
    }

    override fun receiveLongEnergy(maxReceive: Long, simulate: Boolean): Long {
        val received = min(maxReceive,maxLongEnergyStored - longEnergyStored)
        if (!simulate && received != 0L) {
            longEnergyStored += received
            markDirty()
        }
        return received
    }

    override fun extractLongEnergy(maxExtract: Long, simulate: Boolean): Long {
        val extracted = min(maxExtract, longEnergyStored)
        if (!simulate && extracted != 0L) {
            longEnergyStored -= extracted
            markDirty()
        }
        return extracted
    }

    override fun writeBuffer(buffer: FriendlyByteBuf) {
        buffer.writeLong(this.longEnergyStored)
        buffer.writeLong(this.maxLongEnergyStored)
    }

    override fun readBuffer(buffer: FriendlyByteBuf) {
        this.longEnergyStored = buffer.readLong()
        this.maxLongEnergyStored = buffer.readLong()
    }

    override fun onSync() {
        isDirty = false
    }
}